src/persist-client/src/schema.rs

```diff
@@ -76,4 +390,334 @@ mod tests {
+                arrays.push(StringBuilder::new());
+            }
+            Ok(StringsEncoder { fields, arrays })
+        }
+    }
+
+    #[derive(Debug)]
+    struct StringsDecoder(Vec<StringArray>);
+    impl ColumnDecoder<Strings> for StringsDecoder {
+        fn decode(&self, idx: usize, val: &mut Strings) {
+            val.0.clear();
+            for col in self.0.iter() {
+                if col.is_valid(idx) {
+                    val.0.push(col.value(idx).into());
+                } else {
+                    val.0.push("".into());
+                }
+            }
+        }
+        fn is_null(&self, _: usize) -> bool {
+            false
+        }
+        fn stats(&self) -> StructStats {
+            StructStats {
+                len: self.0[0].len(),
+                cols: Default::default(),
+            }
+        }
+    }
+
+    #[derive(Debug)]
+    struct StringsEncoder {
+        fields: Vec<Field>,
+        arrays: Vec<StringBuilder>,
+    }
+    impl ColumnEncoder<Strings> for StringsEncoder {
+        type FinishedColumn = StructArray;
+
+        fn append(&mut self, val: &Strings) {
+            for (idx, val) in val.0.iter().enumerate() {
+                if val.is_empty() {
+                    self.arrays[idx].append_null();
+                } else {
+                    self.arrays[idx].append_value(val);
+                }
+            }
+        }
+        fn append_null(&mut self) {
+            unreachable!()
+        }
+        fn finish(self) -> Self::FinishedColumn {
+            let arrays = self
+                .arrays
+                .into_iter()
+                .map(|mut x| ArrayBuilder::finish(&mut x))
+                .collect();
+            StructArray::new(self.fields.into(), arrays, None)
+        }
+    }
+
+    #[mz_persist_proc::test(tokio::test)]
+    #[cfg_attr(miri, ignore)]
+    async fn compare_and_evolve_schema(mut dyncfgs: ConfigUpdates) {
+        dyncfgs.add(&DANGEROUS_ENABLE_SCHEMA_EVOLUTION, true);
+
+        let client = new_test_client(&dyncfgs).await;
+        let d = Diagnostics::for_tests();
+        let shard_id = ShardId::new();
+        let schema0 = StringsSchema(vec![false]);
+        let schema1 = StringsSchema(vec![false, true]);
+
+        let write0 = client
+            .open_writer::<Strings, (), u64, i64>(
+                shard_id,
+                Arc::new(schema0.clone()),
+                Arc::new(UnitSchema),
+                d.clone(),
+            )
+            .await
+            .unwrap();
+        assert_eq!(write0.write_schemas.id.unwrap(), SchemaId(0));
+
+        // Not backward compatible
+        let res = client
+            .compare_and_evolve_schema::<Strings, (), u64, i64>(
+                shard_id,
+                SchemaId(0),
+                &StringsSchema(vec![]),
+                &UnitSchema,
+                d.clone(),
+            )
+            .await
+            .unwrap();
+        assert_eq!(res, CaESchema::Incompatible);
```

Comment:
In this case are we just dropping the one column? Which should be compatible in the future but isn't now? Apologies if I'm missing something!